pkg/kv/kvserver/store_create_replica.go

```diff
@@ -106,64 +167,40 @@ func (s *Store) tryGetOrCreateReplica(
 	creatingReplica *roachpb.ReplicaDescriptor,
 ) (_ *Replica, created bool, _ error) {
 	// The common case: look up an existing (initialized) replica.
-	if repl, ok := s.mu.replicasByRangeID.Load(rangeID); ok {
-		repl.raftMu.Lock() // not unlocked on success
-		repl.mu.RLock()
-
-		// The current replica is removed, go back around.
-		if repl.mu.destroyStatus.Removed() {
-			repl.mu.RUnlock()
-			repl.raftMu.Unlock()
-			return nil, false, errRetry
-		}
+	if repl, err := s.tryGetReplica(ctx, rangeID, replicaID, creatingReplica); err != nil {
+		return nil, false, err
+	} else if repl != nil {
+		return repl, false, nil
+	}
 
-		// Drop messages from replicas we know to be too old.
-		if fromReplicaIsTooOldRLocked(repl, creatingReplica) {
-			repl.mu.RUnlock()
-			repl.raftMu.Unlock()
-			return nil, false, roachpb.NewReplicaTooOldError(creatingReplica.ReplicaID)
-		}
+	// No replica currently exists, so we'll try to create one. Multiple
+	// goroutines may be racing at this point, so grab a "lock" over this rangeID
+	// in s.creatingReplicas map for one goroutine, and retry others.
+	s.creatingReplicas.Lock()
+	defer s.creatingReplicas.Unlock()
+	if _, ok := s.creatingReplicas.m[rangeID]; ok {
+		return nil, false, errRetry
```

Comment:
// Lost the race - another goroutine is currently creating that replica. Let the caller retry
// so that they can eventually see it.